src/redash_stmo/datasource_link.py

```diff
@@ -0,0 +1,68 @@
+    "Cassandra": "http://cassandra.apache.org/doc/latest/cql/index.html",
+    "dynamodb_sql": "https://dql.readthedocs.io/en/latest/",
+    "baseelasticsearch": "https://www.elastic.co/guide/en/elasticsearch/reference/current/index.html",
+    "google_spreadsheets": "http://redash.readthedocs.io/en/latest/datasources.html#google-spreadsheets",
+    "hive": "https://cwiki.apache.org/confluence/display/Hive/LanguageManual",
+    "impala": "http://www.cloudera.com/documentation/enterprise/latest/topics/impala_langref.html",
+    "influxdb": "https://docs.influxdata.com/influxdb/v1.0/query_language/spec/",
+    "jirajql": "https://confluence.atlassian.com/jirasoftwarecloud/advanced-searching-764478330.html",
+    "mongodb": "https://docs.mongodb.com/manual/reference/operator/query/",
+    "mssql": "https://msdn.microsoft.com/en-us/library/bb510741.aspx",
+    "mysql": "https://dev.mysql.com/doc/refman/5.7/en/",
+    "oracle": "http://docs.oracle.com/database/121/SQLRF/toc.htm",
+    "pg": "https://www.postgresql.org/docs/current/",
+    "redshift": "http://docs.aws.amazon.com/redshift/latest/dg/cm_chap_SQLCommandRef.html",
+    "presto": "https://prestodb.io/docs/current/",
+    "python": "http://redash.readthedocs.io/en/latest/datasources.html#python",
+    "insecure_script": "http://redash.readthedocs.io/en/latest/datasources.html#python",
+    "sqlite": "http://sqlite.org/lang.html",
+    "treasuredata": "https://docs.treasuredata.com/categories/hive",
+    "url": "http://redash.readthedocs.io/en/latest/datasources.html#url",
+    "vertica": (
+        "https://my.vertica.com/docs/8.0.x/HTML/index.htm#Authoring/"
+        "ConceptsGuide/Other/SQLOverview.htm%3FTocPath%3DSQL"
+        "%2520Reference%2520Manual%7C_____1"
+    )
+}
+
+
+class DataSourceLinkResource(BaseResource):
+    def get(self, data_source_id):
+        data_source = get_object_or_404(
+            DataSource.get_by_id_and_org,
+            data_source_id,
+            self.current_org,
+        )
+        require_access(data_source.groups, self.current_user, view_only)
+        try:
+            result = {
+                "type_name": data_source.query_runner.name(),
+                "doc_url": data_source.options.get("doc_url", None)
+            }
+        except Exception as e:
+            return {"message": unicode(e), "ok": False}
+        else:
+            return {"message": result, "ok": True}
+
+def datasource_link(app=None):
+    for runner_type, runner_class in query_runners.items():
+        if runner_type not in DATASOURCE_URLS:
+            continue
+
+        runner_class.add_configuration_property("doc_url", {
+            "type": "string",
+            "title": "Documentation URL",
+            "default": DATASOURCE_URLS[runner_type]})
+
+    # After api.init_app() is called, api.app should be set by Flask (but it's not) so that
+    # further calls to add_resource() are handled immediately for the given app.
+    api.app = app
```

Comment:
Is this an issue in Redash or something that Flask-Restful does wrong?

Reply:
This is an issue in Flask-Restful. There are a few things going on that led me to adding this line.

tl;dr: Flask-Restful does not play well with the Redash setup of multiple workers each having a different instance of a flask app.

1. Inside `create_app()` in `redash/__init__.py`, we cannot move `handlers.init_app(app)` (which calls `api.init_app(app)`) after `extensions.init_extensions(app)` because the datasource_health extension overwrites the `status_api` endpoint after it’s created. If extensions initialize first then an attempt is made in Redash to add `status_api` and it crashes. So we keep the init order so redash-stmo can override endpoints that already exist.

2. Unlike `api = Api(app)` (https://github.com/flask-restful/flask-restful/blob/master/flask_restful/__init__.py#L92), `api.init_app()` does not set `api.app` when it's called (https://github.com/flask-restful/flask-restful/blob/master/flask_restful/__init__.py#L196). This results in any further calls to `api.add_resource()` after `api.init_app()` to actually have no effect. It results in appending to `api.resources` (https://github.com/flask-restful/flask-restful/blob/master/flask_restful/__init__.py#L384) but no call to `register_view()`

3. Even if we did manage to work around this issue in `1.`, this would not resolve this issue. A Flask app is created in each worker with `create_app()`. If an extension tries to do `api.add_org_resource()`, the same resource will be added to `api.resources` from multiple workers. Now when `api.init_app(app)` is called, there is an exception thrown that there are multiple view functions with the same name.

The crux of the problem is that `api.resources` is shared among multiple apps as a temporary cache until `api.init_app(app)` is called so that resources can be attached to a given app. Since we have a different app for each worker, and each worker is adding the same resource in an extension, this model doesn't quite work.

By setting `api.app = app` we ensure that the resource is added to the correct app and there is no duplication by caching in api.resources.